Update webhook-request-parser.md - Round #1 (#25)
SymfonyCasts/lemon-squeezy: sfcasts/webhook-request-parser.md

# Implementing the Webhook Request Parser

In the last chapter, we saw LemonSqueezy deliver its first webhook to our local website via the Ngrok tunnel. *But* it didn't *quite* hit the mark because we still need to set up the actual webhook handling. So, let's roll up our sleeves and get to work!

When we ran the `make:webhook` command, it generated two files - a request parser and a consumer. Let's start with the parser. Open up `src/Webhook/LemonSqueezyRequestParser.php`.

It's important to remember that *all* LemonSqueezy webhooks are sent as a POST request with a JSON payload to the `/webhook/lemon-squeezy` path we set up in the dashboard. With *that* in mind, we need to tweak the `getRequestMatcher()` method.

> Inside `getRequestMatcher()`, we have `ChainRequestMatcher`. This is basically a combination of these three matchers. Set the `PathRequestMatcher` to `/webhook/lemon-squeezy`. Then we need to confirm that it's a `POST` request. To keep things simple, use the `Request::METHOD_POST` constant, which is just a fancy way of saying "POST". This `isJsonRequestMatcher` is fine as is and doesn't need any arguments.

Reviewer: Maybe this? Not sure though it's better

```suggestion
Inside `getRequestMatcher()`, we have `ChainRequestMatcher`. This basically includes the other three matchers. Set the `PathRequestMatcher` to `/webhook/lemon-squeezy`. Then we need to confirm that it's a `POST` request. To make it cooler, use the `Request::METHOD_POST` constant, which is just a fancy way of saying "POST". This `isJsonRequestMatcher` is fine as is and doesn't need any arguments.
```

Okay, we're ready to move on to the next method: `doParse()`. The first order of business here is to verify the webhook signature. To keep things neat, we'll create a separate method for this. Add `private function verifySignature()`... and this function will take *two* arguments: `Request $request` and `string $secret`. The return type is `void`.

Next, we have to calculate the request payload's hash using LemonSqueezy's algorithm: `$payload = $request->getContent()`. We'll start with a hash variable: `$hash = hash_hmac('sha256', $payload, $secret)`. Then we need to fetch the signature from the request header with `$signature = $request->headers->get('X-Signature', '')`.

Now it's time to see if the hash matches the signature. If it *does* (`if (hash_equals($hash, $signature))`), then we're good to go. If *not*, we'll `throw new RejectWebhookException()` with a 401 `Response::HTTP_UNAUTHORIZED` status code. We'll also add a message - `Invalid LemonSqueezy signature!` - so we know what happened.

> Once that's done, we can call this from the `doParse()` method at the beginning. I'll de-clutter and remove this placeholder code, and *then* we'll validate the payload: `$payload = $request->toArray()`. Below, say `$eventName = $payload['meta']['event_name']` and `$webhookId =$payload['meta']['webhook_id']`.

Reviewer: ```suggestion
Once that's done, we can call this from the `doParse()` method at the beginning. I'll de-clutter and remove this placeholder code, and *then* we'll validate the payload: `$payload = $request->toArray()`. Below, say `$eventName = $payload['meta']['event_name']` and `$webhookId = $payload['meta']['webhook_id']`.
```
